lib/task.js

```diff
@@ -238,141 +209,100 @@ function factory(
         assert.string(this.definition.runJob, 'Task.definition.runJob injector string');
         // This should already have been validated to exist
         var Job = injector.get(this.definition.runJob);
-        this.job = new Job(this.options, this.parentContext, this.instanceId);
+        this.job = new Job(this.options, this.context, this.instanceId);
         assert.func(this.job.run, "Task Job run method");
         assert.func(this.job.cancel, "Task Job cancel method");
     };
 
-    Task.prototype.publishFinished = function() {
-        var self = this;
-        return eventsProtocol.publishTaskFinished(self.instanceId)
-        .catch(function(error) {
-            logger.error("Error publishing Task finished event.", {
-                error: error,
-                taskId: self.instanceId,
-                taskName: self.name,
-                job: self.definition.runJob
-            });
-        });
-    };
-
-    Task.prototype._run = function () {
+    Task.prototype.run = function() {
         var self = this;
-
-        logger.info("Running task job.", {
-            taskId: self.instanceId,
-            name: self.definition.friendlyName,
-            job: self.definition.runJob
-        });
-
-        // return for test
-        return self.job.run()
+        if (self.state !== TaskStates.Pending) {
+            return Promise.resolve();
+        }
+        self.state = TaskStates.Running;
+        return Promise.resolve()
+        .then(function() {
+            return self._run();
+        })
         .then(function() {
             self.state = TaskStates.Succeeded;
         })
+        .then(function() {
+            return self;
+        })
         .catch(Errors.TaskCancellationError, function(e) {
             self.state = TaskStates.Cancelled;
             self.error = e;
+            return self;
+        })
+        .catch(Errors.TaskStopError, function(e) {
+            self.state = TaskStates.Pending;
+            self.error = e;
+            return null;
         })
         .catch(Errors.TaskTimeoutError, function(e) {
             self.state = TaskStates.Timeout;
             self.error = e;
+            return self;
         })
         .catch(function(e) {
             self.state = TaskStates.Failed;
             self.error = e;
-        })
-        .finally(function() {
-            return self.finish();
+            return self;
         });
     };
 
-    Task.prototype.run = function() {
-        if (this.state === TaskStates.Running) {
-            return Promise.resolve();
-        }
-        this.state = TaskStates.Running;
-
-        try {
-            this.renderOwnOptions(this.options);
-            this.instantiateJob();
-        } catch (e) {
-            this.state = TaskStates.Failed;
-            this.error = e;
-            return this.finish();
-        }
-
-        return this._run();
-    };
-
-    Task.prototype.finish = function() {
+    Task.prototype._run = function() {
```

Comment:
Outer `run` calls inner `_run` with a bunch of error handling conditionals and so forth. It was more readable to split these out.